### DISCUSSION TITLE
Setting up on MAcOSX for development.

### DISCUSSION THREAD

**Discussion by oxoocoffee** [NanoVNA-Saver/nanovna-saver#620] [General]
Hello,

I am trying to set up my machine to help fix some bugs that I do encounter on MacOSX myself. I think I got most of the setup done. I can start the debugger and hit the first break point at `nanovna-server.py`.
```
if __name__ == "__main__":
    main()
```

The issue is that when I try to step-in, the application just continues "over" and I see the main screen right away. I did open `src/__main__.py` and set up a breakpoint on the first line but it never stops on it.
```
def main():
    parser = argparse.ArgumentParser(
```

Can someone help me figure out what am I missing or doing wrong?

Here is a little about my setup/config:

MacOSX Venture running on M1
Python 3.11.2 installed with homebrew
IDE is Visual Studio Code
virtual env created and sourced in
Here is list of pip packages installed

```
Package           Version
----------------- ------------------------------------
Cython            0.29.33
NanoVNASaver      0.6.0.post1.dev3+g52cdac4f.d20230316
numpy             1.24.2
packaging         23.0
pip               23.0.1
PyQt6             6.4.2
PyQt6-Qt6         6.4.2
PyQt6-sip         13.4.1
pyserial          3.5
scipy             1.10.1
setuptools        67.6.0
setuptools-scm    7.1.0
typing_extensions 4.5.0
wheel             0.40.0
```

Here is a .vscode/launch.json contents

```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Python: Current File",
            "type": "python",
            "request": "launch",
            "program": "${file}",
            "console": "integratedTerminal",
            "justMyCode": true
        }
    ]
}
```

A side note: Reading CONTRIBUTING.rst I did followed information but found some issues/inconsistency. 
** You should run:: ** section states:

```
pip install -U pip setuptools -e .
```

but I had to run this command from setup.py in other to make it work

```
pip install -U setuptools setuptools_scm wheel
```

Also, I had to manually run from the project root directory. This was not mentioned. 

```
pip install .
```

Please note all pip commands were executed after sourcing venv

**Comment by ikatkov** [comment]
When you run `pip install .` in the terminal, you are installing a Python package that is located in the current working directory. The `.` refers to the current directory.

The location where the files are installed depends on the operating system and the Python installation. Generally, the packages are installed in the site-packages directory of the Python installation.

On Linux and macOS systems, the site-packages directory is typically located in /usr/local/lib/pythonX.X/site-packages/, where X.X represents the Python version number.

This command reads the setup.py file in the current directory and installs the package based on the information provided in that file. The setup.py file typically includes information such as the package name, version, author, dependencies, and other metadata.

The breakpoint does not work, because it does not run the code that you have open in the editor, it runs code in `/usr/local/lib/pythonX.X/site-packages/*`

**Reply by oxoocoffee** [reply]
@ikatkov I am pretty sure that after 30 years of c/c++ on different platforms, I think I know what “.” is :) 
Perhaps python is your primary language, which would explain why you think, if I can call it “common sense” and perhaps this made you focus on “.” not  understanding it.
But what was not clear to “me”, as I am doing some python here and there, that you have to run “pip install”. Had to gold few times.
So my point was, I was trying to point out, to add this to documentation for folks who are not fluent  in python, but are interested and like to donate their time to a project.
I hope you understand my point :)

By the way, after your explanation, and pointing out where the problem is, you did not explain what specifically to do or where to look so I can check and fix or report. So in a way I am still at the same point as I was before your post.

Cheers

**Comment by dothebart** [comment]
On OSX you better not lean on the shebang to pick the proper python. Even more so on the ARM machines. You may end up it installing you x86 .so to you.

**Comment by Ho-Ro** [comment]
My experience (on a Linux system) is that you do not need a lot of installation, I did a simple clone of the GH repo and started the saver by `python nanovna-saver.py` just from the top directory. In the beginning some modules were missing that I was able to install once via `apt install pytho3-xyz...` until the program came up. As I'm on Debian I prefer `apt` over `pip`.
Then I could e.g. simply do a new `git clone ...` into a different directory and run the program from there.

This procedure is fully ok for me doing development on nanovna-saver.

Maybe off-topic?

**Reply by dothebart** [reply]
yes, this question is better asked to the specific python vendor - homebrew in this case!

**Reply by oxoocoffee** [reply]
@Ho-Ro 
Tx for input. Are you debugging using 
1. python/print
2. pdb
3. or some IDE? In me case Visual Studio Code?

Reading more, I think I am missing at list first entry below in my json config file

```
   “cwd": "${workspaceFolder}",
   "args": ["-f", "input_filename"]
```
Will try it a bit later today

**Reply by Ho-Ro** [reply]
> Are you debugging ...

As I do mostly GUI changes (adopt window size, reorder buttons, add scroll bars etc.) I look at the visual result and use some old-school `print(..)` spread into the code.

**Reply by zarath** [reply]
I debug with vscode via debugpy module. See debug.sh in root of repo.
Expect not being able to that breakpoints in the sweeper thread.

**Comment by oxoocoffee** [comment]
By the way, I have other self contained small python project on same machine and using same version of python and IDI and everything works. So it must be something specific here :) I will keep looking

**Comment by oxoocoffee** [comment]
Ok, I was able to figure it out. Here is what I did and the last step was what I needed to adjust. I am running python version `/opt/homebrew/Cellar/python@3.11/3.11.2_1`

From Terminal:
1. Clone the repository to `~/projects/nanovna-saver`
2. Create virtual env in `~/.virtual_py_env/nanovna-saver`
3. source it ` source ~/.virtual_py_env/nanovna-saver/bin/activate`
4. `cd ~/projects/nanovna-saver`
5. Start Visual Studio Code `code .` from the terminal to make sure venv is active.
6. Create launch.json in `.vscode/launch.json` in `~/projects/nanovna-saver`

Here is the content of launch.json

```json
{
  // Use IntelliSense to learn about possible attributes.
  // Hover to view descriptions of existing attributes.
  // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
  "version": "0.2.0",
  "configurations": [
    {
        "name": "Python: main (Integrated Terminal)",
        "type": "python",
        "request": "launch",
        "program": "${workspaceFolder}/nanovna-saver.py",
        "console": "integratedTerminal",
        "justMyCode": false,
        "cwd": "${workspaceFolder}",
        //"args": [""] // ["-f", "file"]
    }
}
```

So the issue was that Visual Studio Code when it created `.vscode/launch.json` file, by default did set `"justMyCode": true`
I had to change it to `false`. Everything started to work.

For others. Here is my `pip list` of packages installed in a virtual environment.

```bash
(nanovna-saver)  ➜ pip list
Package           Version
----------------- ------------------------------------
Cython            0.29.33
NanoVNASaver      0.6.0.post1.dev3+g52cdac4f.d20230316
numpy             1.24.2
packaging         23.0
pip               23.0.1
PyQt6             6.4.2
PyQt6-Qt6         6.4.2
PyQt6-sip         13.4.1
pyserial          3.5
scipy             1.10.1
setuptools        67.6.0
setuptools-scm    7.1.0
typing_extensions 4.5.0
wheel             0.40.0
```

So cool, now I can debug. If there are any arey other gothas or watch out from other developers, please let me know.
